File: universidad/agra/parcial1/zlatan.cpp
/*
Solution Parcial Práctico Opcional (Zlatan)
Daniel Alejandro Posada Noguera
Árboles y Grafos (AGRA)
*/

#include <vector>
#include <string>
#include <iostream>

using namespace std;

int number_from_digits(vector<int> &digits){
    int i, n = digits.size(), number = 0;
    vector<int> pows = {1000, 100, 10, 1};
    for(i = 0; i < n; i++){
        number += digits[i] * pows[i];
    }
}

int bfs_Aux(vector<int> &init, vector<int> &targ, vector<int> &vis){

}

int solve(vector<int> &init, vector<int> &targ, vector<int> &forbid){

}

int convert_to_number(string &s){

}

int main(){
    int cases, aux, i, j, n, ans;
    vector<int> initial(4), target(4), forbiddens;
    cin >> cases;
    for(i = 0; i < cases; i++){
        for(j = 0; j < 4; j++) 
            cin >> initial[j];
        for(j = 0; j < 4; j++)
            cin >> target[j];
        cin >> n;
        forbiddens = vector<int> (n);
        for(i = 0; i < n; i++){
            cin >> aux;
            forbiddens[i] = convert_to_number(aux);
        }
        ans = solve(initial, target, forbiddens);
        cout << ans << endl;
        cin >> aux;
    }

    return 0;
}

File: universidad/agra/parcial1/zlatan.py
'''
Solution Parcial Práctico Opcional (Zlatan)
Daniel Alejandro Posada Noguera
Árboles y Grafos (AGRA)
'''
from sys import stdin
from collections import deque

def number_from_digits(digits):
    number = 0
    for i in range(len(digits)):
        number += digits[i] * pow(10, len(digits) - 1 - i)
    return number

def bfs_aux(initial, target, vis):
    q = deque()
    changes = [-1, 1]
    flag, d_answer = False, -1
    q.append((initial, 0))
    vis[number_from_digits(initial)] = 1
    
    while len(q) > 0 and not flag:
        num, distance = q.popleft()
        if target == num: 
            flag, d_answer = True, distance
        else:
            for i in range(len(num)):
                for c in changes:
                    aux = num[:]
                    aux[i] = (aux[i] + c) % 10
                    val = number_from_digits(aux)
                    if not vis[val]:
                        q.append((aux, distance + 1))
                        vis[val] = 1          
    return d_answer

def solve(initial, target, forbidenns):
    # 0 : no visitado, 1 : visitado, 2 : prohibido
    vis = [0 for _ in range(10000)]
    for i in range(len(forbidenns)):
        vis[forbidenns[i]] = 2
        
    ans = bfs_aux(initial, target, vis)
    return ans
    

def main():
    cases = int(stdin.readline())
    for _ in range(cases):
        initial = list(map(int, stdin.readline().split()))
        target = list(map(int, stdin.readline().split()))
        forbiddens = []
        n = int(stdin.readline())
        for _ in range(n):
            aux = number_from_digits(list(map(int, stdin.readline().split())))
            forbiddens.append(aux)
            
        ans = solve(initial, target, forbiddens)
        print(ans)
        
        _ = stdin.readline()
    
    
main()